route/kafkamdm.go

```diff
@@ -0,0 +1,240 @@
+	tickFlushSize     metrics.Histogram // only updated after successful flush
+	manuFlushSize     metrics.Histogram // only updated after successful flush. not implemented yet
+	numBuffered       metrics.Gauge
+}
+
+// NewKafkaMdm creates a special route that writes to a grafana.net datastore
+// We will automatically run the route and the destination
+func NewKafkaMdm(key, prefix, sub, regex, broker, topic, codec, schemasFile, partitionBy string, bufSize, orgId, flushMaxNum, flushMaxWait, timeout int) (Route, error) {
+	m, err := matcher.New(prefix, sub, regex)
+	if err != nil {
+		return nil, err
+	}
+	schemas, err := getSchemas(schemasFile)
+	if err != nil {
+		return nil, err
+	}
+
+	cleanAddr := util.AddrToPath(broker)
+
+	r := &KafkaMdm{
+		baseRoute: baseRoute{sync.Mutex{}, atomic.Value{}, key},
+		topic:     topic,
+		broker:    broker,
+		buf:       make(chan []byte, bufSize),
+		schemas:   schemas,
+		orgId:     orgId,
+
+		bufSize:      bufSize,
+		flushMaxNum:  flushMaxNum,
+		flushMaxWait: time.Duration(flushMaxWait) * time.Millisecond,
+
+		numErrFlush:       stats.Counter("dest=" + cleanAddr + ".unit=Err.type=flush"),
+		numOut:            stats.Counter("dest=" + cleanAddr + ".unit=Metric.direction=out"),
+		durationTickFlush: stats.Timer("dest=" + cleanAddr + ".what=durationFlush.type=ticker"),
+		durationManuFlush: stats.Timer("dest=" + cleanAddr + ".what=durationFlush.type=manual"),
+		tickFlushSize:     stats.Histogram("dest=" + cleanAddr + ".unit=B.what=FlushSize.type=ticker"),
+		manuFlushSize:     stats.Histogram("dest=" + cleanAddr + ".unit=B.what=FlushSize.type=manual"),
+		numBuffered:       stats.Gauge("dest=" + cleanAddr + ".unit=Metric.what=numBuffered"),
+	}
+
+	r.partitioner, err = partitioner.NewKafka(partitionBy)
+	if err != nil {
+		log.Fatal(4, "kafkaMdm %q: failed to initialize partitioner. %s", r.Key, err)
+	}
+
+	// We are looking for strong consistency semantics.
+	// Because we don't change the flush settings, sarama will try to produce messages
+	// as fast as possible to keep latency low.
+	config := sarama.NewConfig()
+	config.Producer.RequiredAcks = sarama.WaitForAll // Wait for all in-sync replicas to ack the message
+	config.Producer.Retry.Max = 10                   // Retry up to 10 times to produce the message
+	config.Producer.Compression, err = getCompression(codec)
+	if err != nil {
+		log.Fatal(5, "kafkaMdm %q: %s", r.key, err)
+	}
+	config.Producer.Return.Successes = true
+	config.Producer.Timeout = time.Duration(timeout) * time.Millisecond
+	err = config.Validate()
+	if err != nil {
+		log.Fatal(4, "kafkaMdm %q: failed to validate kafka config. %s", r.Key, err)
+	}
+	r.saramaCfg = config
+
+	r.config.Store(baseConfig{*m, make([]*dest.Destination, 0)})
+	go r.run()
+	return r, nil
+}
+
+func (r *KafkaMdm) run() {
+	metrics := make([]*schema.MetricData, 0, r.flushMaxNum)
+	ticker := time.NewTicker(r.flushMaxWait)
+	brokers := []string{r.broker}
+
+	connect := func() {
```

Comment:
I dont understand why are not just initializing the producer directly.  It will be initialized when the first flush occurs or when ticker fires after flushMaxWait.

the producer is never unset, and so will never need to be re-initialized